docs/dep/dep-004.md

```diff
@@ -0,0 +1,146 @@
+preferrable if we could keep using Helm, but in a more appropriate way. At the
+moment we are tightly coupling the definition of a workflow with the deployment
+of its resources. Specifically, we tend to produce a manifest that contains
+both the Tasks AND the Pipeline definition. However, Tasks and Pipelines are
+fundamentally different. Tasks contain container image URLs, for example, or
+they could have various information related to the way that secrets and volumes
+or limits are applied in the cluster where the workflow is expected to run. A
+Pipeline definition should be the exact same between different cluster/
+environments/organisations since it's just a combination of Tasks to produce a
+result.
+
+So our goal should be to modify and extend our code to have a clearer
+separation between the operations and phases of the lifetime of the system.
+
+The first stage of the lifetime of the system consists of the deployment of
+Tasks on the clusterand that should be achieved through Helm packages, allowing
+the system operator to deploy and modify the configuration in whatever way they
+deem necessary.
+
+An example of why we need to do this is a security teams who wishes to have a
+custom base image forall their Tasks, which requires them to build all the
+images from scratch and push them into aprivate container registry. The host of
+the registry is going to be supplied as a parameter to Helm during deployment,
+but it should not be required when deploying a Pipeline. A Pipeline that
+discovers with a high probability vulnerabilities in a Javascript codebase
+should be able to run inany cluster as long as the Tasks are available.
+Packaging all the Tasks into one Helm packagemakes this trivial to achieve,
+since it can be easily uploaded and distributed to an OCI registry.
+
+The second stage begins as soon as the `draconctl` is invoked to deploy a
+Pipeline. The Pipeline lists a set of components. The user has 3 different
+formats to describe a component:
+
+1. a local filesystem path
+
+This is used by the system to, on the fly read a Task and deploy it on the
+cluster. There are not going to be any checks if the Task already exists, or if
+the version of the Task is newer than the Task being fetched. It's going to be
+`applied` on the cluster, as if the user is invoking `kubectl apply ...`.
+
+2. a URL
+
+The URL is going to be resolved and the Task is going to be fetched and
+deployed on the cluster similar to the case where we have a file in the local
+filesystem.
+
+3. a pURL-inspired id of a Task deployed using Helm
+
+As mentioned previously, we wish to package all our Tasks in one Helm package
+that can be easily deployed on a cluster and provides access to all our
+components. When a user wants to reference a Task deployed via Helm on the
+cluster in a Pipeline description, they will have to use a new URI type that is
+inspired by `pURL` and looks like this: `helm://{{ helm_release_name }}/name`.
```

Comment:
nit: 
why not make it `purl://helm/{{ helm_release_name }}/{{name}}` ? then we can use the go-purl library to parse it and we don't need to do any custom magic

Reply:
:thinking: I think you are right that I'm not following purl closely but what you are suggesting is a bit closer but not enough. looking at the [examples](https://github.com/package-url/purl-spec?tab=readme-ov-file#some-purl-examples) once again I think the closest we could go is: `pkg:helm/{{ helm_release_name }}/name` or we could go further and do `pkg:helm/{{ helm_registry }}/{{ package_name }}/name` so that we can be more explicit. WDYT?
(on a side note, this could be an interesting contribution to the pURL spec itself)

Reply:
yes, absolutely, my aim was to do a valid purl so let's do `pkg:helm/{{ helm_registry }}/{{ package_name }}/name`
where for us it is ALWAYS `pkg:helm/{{ helm_registry }}/dracon/{{ package_name }}` so we can query our components

Reply:
So, I created a toy example to see how the purl library would treat the different format and apparently both of the work.
```go
package main

import (
	"fmt"

	packageurl "github.com/package-url/packageurl-go"
)

func main() {
	taskPurl, err := packageurl.FromString("pkg:helm/helm.github.com/dracon/consumer-aggregator@1.1.0")
	fmt.Printf("%v\n%v", taskPurl, err)
}
```
In this case the `Namespace` will be the `helm.github.com/dracon` part.

I think for now we can use the simple version of the purl `pkg:helm/{{ helm_release_name }}/name` where the `Namespace` will be the Helm release name and will be used as a label selector as described. In the future, if we decide to expand our integration with Helm we could allow `Namespaces` that contain the host of the Helm registry and the explicit package from it, but for the time being I think it would too much complexity for no good reason.